node/core/dispute-coordinator/src/real/spam_slots.rs

```diff
@@ -0,0 +1,119 @@
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::{HashMap, HashSet};
+
+use polkadot_primitives::v1::{CandidateHash, SessionIndex, ValidatorIndex};
+
+use crate::real::LOG_TARGET;
+
+/// Type used for counting potential spam votes.
+type SpamCount = u32;
+
+/// How many unconfirmed disputes a validator is allowed to be a participant in (per session).
+///
+/// Unconfirmed means: Node has not seen the candidate be included on any chain, it has not cast a
+/// vote itself on that dispute, the dispute has not yet reached more than a third of
+/// validator's votes and the including relay chain block has not yet been finalized.
+///
+/// Exact number of `MAX_SPAM_VOTES` is not that important here. It is important that the number is
+/// low enough to not cause resource exhaustion, if multiple validators spend their limits. Also
+/// if things are working properly, this number cannot really be too low either, as all relevant
+/// disputes _should_ have been seen as included my enough validators. (Otherwise the candidate
+/// would not have been available in the first place and could not have been included.) So this is
+/// really just a fallback mechanism if things go terribly wrong.
+const MAX_SPAM_VOTES: u32 = 50;
+
+/// Spam slots for raised disputes concerning unknown candidates.
+pub struct SpamSlots {
+	/// Counts per validator and session.
+	///
+	/// Must not exceed `MAX_SPAM_VOTES`.
+	slots: HashMap<(SessionIndex, ValidatorIndex), SpamCount>,
+
+	/// All unconfirmed candidates we are aware of right now.
+	unconfirmed: UnconfirmedDisputes,
+}
+
+/// Unconfirmed disputes to be passed at initialization.
+pub type UnconfirmedDisputes = HashMap<(SessionIndex, CandidateHash), HashSet<ValidatorIndex>>;
+
+impl SpamSlots {
+	/// Recover `SpamSlots` from state on startup.
+	///
+	/// Initialize based on already existing active disputes.
+	pub fn recover_from_state(unconfirmed_disputes: UnconfirmedDisputes) -> Self {
+		let mut slots: HashMap<(SessionIndex, ValidatorIndex), SpamCount> = HashMap::new();
+		for ((session, _), validators) in unconfirmed_disputes.iter() {
+			for validator in validators {
+				let e = slots.entry((*session, *validator)).or_default();
+				*e += 1;
+				if *e > MAX_SPAM_VOTES {
+					tracing::debug!(
+						target: LOG_TARGET,
+						?session,
+						?validator,
+						count = ?e,
+						"Import exceeded spam slot for validator"
+					);
+				}
+			}
+		}
+
+		Self { slots, unconfirmed: unconfirmed_disputes }
+	}
+
+	/// Add an unconfirmed dispute if free slots are available.
+	pub fn add_unconfirmed(
+		&mut self,
+		session: SessionIndex,
+		candidate: CandidateHash,
+		validator: ValidatorIndex,
+	) -> bool {
+		let c = self.slots.entry((session, validator)).or_default();
+		if *c >= MAX_SPAM_VOTES {
+			return false
+		}
+		*c += 1;
+		let validators = self.unconfirmed.entry((session, candidate)).or_default();
+		validators.insert(validator);
```

Comment:
We should only increment `c` and return false if this is `false`. Otherwise, calling `add_unconfirmed` twice with the same candidate&validator will increment spam slots twice.

Reply:
Good point. Multiple identical imports are also spam, but yeah - pure flood should already be handled properly by distribution.